index.js

```diff
@@ -1,174 +1,162 @@
-      if (tt === 'number' || tt === 'boolean') {
-        // return any immutable types
-        return this._decoded
-      }
-      if (isBinary(this._decoded)) return copyBinary(this._decoded)
-      if (immutableTypes.has(typeof this._decoded) || this._decoded === null) {
-        return this._decoded
-      }
-      return clone(this._decoded)
-    }
+  async equals (block) {
+    if (block === this) return true
+    if (block.asBlock !== block) throw new Error('Not a block instance')
+    const [a, b] = await Promise.all([this.cid(), block.cid()])
+    return a.equals(b)
+  }
+}
 
-    decodeUnsafe () {
-      if (!this._decoded) this._decode()
-      return this._decoded
-    }
+Block.codecs = new Map()
 
-    reader () {
-      return reader(this.decodeUnsafe())
-    }
+Block.add = codec => {
+  if (codec.name) Block.codecs.set(codec.name, codec)
+  if (codec.code) Block.codecs.set(codec.code, codec)
+}
+Block.add(json)
+Block.add(raw)
 
-    async equals (block) {
-      if (block === this) return true
-      const cid = await this.cid()
-      if (CID.asCID(block)) return cid.equals(CID.asCID(block))
-      // https://github.com/bcoe/c8/issues/135
-      /* c8 ignore next */
-      return cid.equals(await block.cid())
-    }
+Block.encoder = (source, codec, hasher = sha256) => {
+  if (typeof codec === 'string') codec = Block.codecs.get(codec)
+  if (!codec) throw new Error('Missing codec')
+  return new Block({ source, codec, hasher })
+}
+Block.decoder = (data, codec, hasher = sha256) => {
+  if (typeof codec === 'string') codec = Block.codecs.get(codec)
+  if (!codec) throw new Error('Missing codec')
+  return new Block({ data, codec, hasher })
+}
+Block.createUnsafe = (data, cid, { hasher, codec } = {}) => {
+  codec = codec || Block.codecs.get(cid.code)
+  if (!codec) throw new Error(`Missing codec ${cid.code}`)
+  return new Block({ data, cid, codec, hasher: hasher || null })
+}
+Block.create = async (data, cid, { hasher, codec } = {}) => {
+  if (typeof cid === 'string') cid = CID.parse(cid)
+  hasher = hasher || Block.codec.get(cid.multihash.code)
+  if (!hasher) {
+    const { code } = cid.multihash.code
+    throw new Error(`Missing hasher for verification. Pass hasher for hash type ${code} or use createUnsafe()`)
   }
-
-  const BlockWithIs = withIs(Block, { className: 'Block', symbolName: '@ipld/block' })
-  BlockWithIs.encoder = (source, codec, algo) => new BlockWithIs({ source, codec, algo })
-  BlockWithIs.decoder = (data, codec, algo) => new BlockWithIs({ data, codec, algo })
-  BlockWithIs.create = (data, cid) => {
-    if (typeof cid === 'string') cid = CID.from(cid)
-    return new BlockWithIs({ data, cid })
+  const hash = await hasher.digest(data)
+  if (!bytes.equals(cid.multihash.bytes, hash.bytes)) {
+    throw new Error('CID hash does not match data')
   }
-  BlockWithIs.multiformats = multiformats
-  BlockWithIs.CID = CID
-  return BlockWithIs
+  return Block.createUnsafe(data, cid, { hasher, codec })
```

Comment:
I think this this is a reason why `createUnsafe` to takes `hasher`. Just using constructor here can simplify `createUnsafe`.

```suggestion
    return new Block({ cid, data, codec, hasher })
```